doc: add shared memory based inter-vm communication introduction (#4865)
projectacrn/acrn-hypervisor: doc/developer-guides/hld/ivshmem-hld.rst

.. _ivshmem-hld:

ACRN Shared Memory Based Inter-VM Communication
###############################################

To support inter virtual machines communication based on shared memory
mechanism, ACRN device model/hypervisor emulate a virtual PCI device

> (Called ivshmem device) to expose the base address and size of shared

Reviewer: ```suggestion
(called ``ivshmem`` device) to expose the base address and size of shared
```

memory.

Inter-VM Communication Overview
*******************************

.. figure:: images/ivshmem-image1.png
   :align: center
   :name: ivshmem-architecture-overview

   ACRN shared memory based inter-vm communication architecture

> Two solutions of ivshmem device: DM-land and hypervisor-land. For DM-land,

Reviewer 2: DM-land -> dm-land

> Two solutions of ivshmem device: dm-land and hypervisor-land. For dm-land,

Reviewer: Are these two solutions two components of a single solution, or are these alternatives to each other?

I'm not a big fan of calling these `dm-land` and `hyervisor-land`, we've not used these in the past. I think I would prefer to use plain English to say if something runs in the Service VM or in the Hypervisor. @dbkinder , thoughts?

Author: > Are these two solutions two components of a single solution, or are these alternatives to each other?

Currently, they are alternatives. dm-land shared memory is allocated from Service VM, hypervisor-land shared memory is allocated from hypervisor.  In the future, when the hypervisor-land is implemented, we will check whether these two solutions can be merged into one solution.

Reviewer: > > Are these two solutions two components of a single solution, or are these alternatives to each other?
> 
> Currently, they are alternatives. dm-land shared memory is allocated from Service VM, hypervisor-land shared memory is allocated from hypervisor. In the future, when the hypervisor-land is implemented, we will check whether these two solutions can be merged into one solution.

Thanks for the clarification!

the ivshmem device emulated in acrn device model and its shared memory region

> is allocated from SOS's memory space. For hypervisor-land, it emulated in

Reviewer: ```suggestion
is allocated from the Service VM's memory space. For hypervisor-land, it is emulated in
```

hypervisor and the shared memory regions are reserved in hypervisor's memory

> space. The former solution only support interactive between port-launched VMs

Reviewer 2: support -> supports

> space. The former solution only supports interactive between port-launched VMs

Reviewer: ```suggestion
space. The former solution only supports interactive between post-launched VMs
```

.. _ivshmem-hld:

ACRN Shared Memory Based Inter-VM Communication
###############################################

To support inter virtual machines communication based on shared memory
mechanism, ACRN device model/hypervisor emulate a virtual PCI device

> (Called ivshmem device) to expose the shared memory base address and

Reviewer 2: to expose the base address and size of shared memory.

size. Currently, ACRN only supports post-launched to post-launched
communication via ivshmem device. In future, ACRN will supports

> pre-launched VM inter-VM communication.

Reviewer 2: Remove "Currently, ACRN only supports post-launched to post-launched communication via ivshmem device. In future, ACRN will supports pre-launched VM inter-VM communication."
We can talk it later, not at the beginning.

Inter-VM Communication Overview
*******************************

.. figure:: images/ivshmem-image1.png
   :align: center
   :name: ivshmem-architecture-overview

>    ACRN Inter-VM Communication Architecture Based on Shared Memory

Reviewer 2: ACRN shared memory based inter-vm communication architecture

Two solutions for ivshmem device implementation, one is called DM-land,
another is called Hypervisor-land. For DM-land, the ivshmem device is
emulated in device model and the inter-VM communication shared memory
region is allocated from Service memory. For Hypervisor-land, the
ivshmem device is emulated in hypervisor and the shared memory is
allocated from hypervisor. The former only work for post-launched VMs,
the latter can work for both of pre-launched and post-launched VMs. The

> Hypervisor-land will be implemented in the future.

Reviewer 2: Two solutions of ivshmem device: DM-land and hypervisor-land.  For DM-land, the ivshmem device emulated in acrn device model and its shared memory region is allocated from SOS's memory space. For hypervisor-land, it emulated in hypervisor and the shared memory regions are reserved in hypervisor's memory space. The former solution only support interactive between port-launched VMs and the latter one can work for both pre-launched and post-launched VMs. The hypervisor-land solution is WIP, plan to support in the near future.

-  Ivshmem dm – The ivshmem device model implements device related
   registers and shared memory mapping. In the future it will also
   support the interrupt mechanism.

> 

Reviewer 2: The ivshmem device model implements register virtualization and shared memory mapping. Will support notification/interrupt mechanism in future.

-  Ivshmem server – A daemon for inter-VM notification capability,
   currently, it has not been implemented, so the inter-VM communication
   doesn’t support notification mechanism.

Ivshmem Device Introduction
***************************

The ivshmem device is a virtual standard PCI device, it consists of two
BARs(BAR0 and BAR2), one is used for interrupt related register and BAR

> size is 256 bytes, another is used for exposing shared memory base

Reviewer 2: one is used for emulate interrupt related registers, another one is used for exposing shared memory region. The ivshmem device doesn't support any extra capabilities.
